examples/wakaama/dtlsconnection.c

```diff
@@ -0,0 +1,570 @@
+   connP = connection_new_incoming(connList, &addr, pnum);
+
+   // do we need to start tinydtls?
+   DEBUG("connection_create(): checking connP\n");
+   if (connP != NULL) {
+      connP->securityObj = securityObj;
+      connP->securityInstId = instanceId;
+      connP->lwm2mH = lwm2mH;
+
+      if (security_get_mode(connP->securityObj, connP->securityInstId) !=
+          LWM2M_SECURITY_MODE_NONE) {
+         connP->dtlsContext = get_dtls_context(connP);
+      } else {
+         // no dtls session
+         connP->noSec = true;
+      }
+   }
+
+   DEBUG("connection_create(): done\n");
+   return connP;
+}
+
+void connection_free(dtls_connection_t *connList) {
+   dtls_free_context(dtlsContext);
+   dtlsContext = NULL;
+   while (connList != NULL) {
+      dtls_connection_t *nextP;
+
+      nextP = connList->next;
+      free(connList);
+
+      connList = nextP;
+   }
+}
+
+int connection_send(dtls_connection_t *connP, uint8_t *buffer, size_t length) {
+   int bw = 0;
+   DEBUG("connection_send(%d)\n", length);
+   if (connP->noSec) {
+      // no security
+      if (0 != send_data(connP, buffer, length)) {
+         return -1;
+      }
+   } else {
+      if (DTLS_NAT_TIMEOUT > 0 &&
+          (lwm2m_gettime() - connP->lastSend) > DTLS_NAT_TIMEOUT) {
+         // we need to rehandhake because our source IP/port probably changed
+         // for
+         // the server
+         if (connection_rehandshake(connP, false) != 0) {
+            printf("can't send due to rehandshake error\n");
+            return -1;
+         }
+      }
+      DEBUG("dtls_write(%d)\n", length);
+      bw = dtls_write(connP->dtlsContext, connP->dtlsSession, buffer, length);
+      if (bw < 0) {
+         DEBUG("dtls_write() failed\n");
+         return -1;
+      }
+      DEBUG("dtls_write() wrote %d bytes\n", bw);
+   }
+
+   return 0;
+}
+
+int connection_handle_packet(dtls_connection_t *connP, uint8_t *buffer,
+                             size_t numBytes) {
+   if (!connP->noSec) {
+      // Let liblwm2m respond to the query depending on the context
+      DEBUG("connection_handle_packet() -> dtls_handle_message()\n");
+      int result = dtls_handle_message(connP->dtlsContext, connP->dtlsSession,
+                                       buffer, numBytes);
+      if (result != 0) {
+         printf("error dtls handling message %d\n", result);
+      }
+      return result;
+   } else {
+      // no security, just give the plaintext buffer to liblwm2m
+      DEBUG("connection_handle_packet() -> lwm2m_handle_packet()\n");
+      lwm2m_handle_packet(connP->lwm2mH, buffer, numBytes, (void *)connP);
+      return 0;
+   }
+}
+
+int connection_rehandshake(dtls_connection_t *connP, bool sendCloseNotify) {
+   // if not a dtls connection we do nothing
+   if (connP->noSec) {
+      return 0;
+   }
+
+   // reset current session
+   dtls_peer_t *peer = dtls_get_peer(connP->dtlsContext, connP->dtlsSession);
```

Comment:
It's not less aggressive in resources to use `dtls_renegotiate()`  ?

Or maybe not, as you truly want to terminate  the relation between the peers.

Reply:
This is code I've taken as is from the Wakaama example, so I'm not entirely sure which of the two should be used here.

Reply:
Ok. For purposes of comparing performance it will be better not touching it then. 

Unless that another reviewer has a different opinion.